ksqldb-engine/src/main/java/io/confluent/ksql/security/KsqlAuthorizationValidatorImpl.java

```diff
@@ -150,12 +166,51 @@ private String getSourceTopicName(final MetaStore metaStore, final SourceName st
     return dataSource.getKafkaTopicName();
   }
 
-  private String getCreateAsSelectSinkTopic(
+  private KsqlTopic getCreateAsSelectSinkTopic(
       final MetaStore metaStore,
       final CreateAsSelect createAsSelect
   ) {
-    return createAsSelect.getProperties().getKafkaTopic()
-        .orElseGet(() -> getSourceTopicName(metaStore, createAsSelect.getName()));
+    final CreateSourceAsProperties properties = createAsSelect.getProperties();
+
+    final String sinkTopicName;
+    final KeyFormat sinkKeyFormat;
+    final ValueFormat sinkValueFormat;
+
+    if (!properties.getKafkaTopic().isPresent()) {
+      final DataSource dataSource = metaStore.getSource(createAsSelect.getName());
+      if (dataSource != null) {
+        sinkTopicName = dataSource.getKafkaTopicName();
+        sinkKeyFormat = dataSource.getKsqlTopic().getKeyFormat();
+        sinkValueFormat = dataSource.getKsqlTopic().getValueFormat();
+      } else {
+        throw new KsqlException("Cannot validate for topic access from an unknown stream/table: "
+            + createAsSelect.getName());
+      }
+    } else {
+      final Format keyFormat = FormatFactory.fromName(
+          properties.getKeyFormat().orElse(defaultTopicKeyFormat));
+      final Format valueFormat = FormatFactory.fromName(
+          properties.getValueFormat().orElse(defaultTopicValueFormat));
+
+      sinkTopicName = properties.getKafkaTopic().get();
+
+      sinkKeyFormat = KeyFormat.of(
+          FormatInfo.of(keyFormat.name()),
+          keyFormat.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)
+              ? SerdeFeatures.of(SerdeFeature.SCHEMA_INFERENCE)
+              : SerdeFeatures.of(),
```

Comment:
`keyFormat` has a `supportedFeatures()` method that returns a set of features. I could call `SerdeFeatures.from(keyFormat.supportedFeatures())` here, but `SerdeFeatures.from` throws an exception because of incompatible features returned. Because I just need SCHEMA_INFERENCE, then I just added this condition here.

Reply:
IIUC, the only way to hit the incompatible features exception is if UNWRAP_SINGLES and WRAP_SINGLES are both included. Why is that happening?